### ISSUE TITLE
Unmounted parts of the form are not being validated

### ISSUE THREAD

**Issue by dziamid** [erikras/redux-form#2321]
I have a form where only some parts of the form are visible at any given time, think  about an accordion component where only one tab is mounted at any give time. 

If you submit such form, it will call onSubmit prop with data from all parts of the form (including the once that are unmounted atm). However, if you happen to have validation errors in unmounted parts of the form, they will simply be ignored. and onSubmit gets called with invalid data!

**Comment by 0rangeT1ger** [comment]
Maybe it's feature, not bug.
Come to think of it, it is kind of a big and useful feature.

**Comment by goshakkk** [comment]
Are you using field-level validations? If so, consider changing to form-level validations instead: https://github.com/erikras/redux-form/issues/2451

**Comment by gustavohenke** [comment]
Hi @dziamid, can you please give us a reproducible example?

**Comment by dziamid** [comment]
@gustavohenke sure, do you guys have a working redux-form fiddle to fork from?

**Comment by dnzyt** [comment]
I'm experiencing the same issue. The validation doesn't happen on the unmounted field.

**Comment by gustavohenke** [comment]
Hi @dziamid, sure: https://jsfiddle.net/gustavohenke/48814zsq/

**Comment by dziamid** [comment]
@gustavohenke he you go https://jsfiddle.net/hw1kqamL/

**Comment by gustavohenke** [comment]
So, the field is unregistered when unmounted. 
This means Redux Form is forgiving the error for such field.

Shall we introduce yet another config to control this behavior, @erikras?

**Comment by dziamid** [comment]
Would be great to define some kind of Form schema that would register all fields and define basic validation rules that can be further reinforced by validate function.

**Comment by psimoneau22** [comment]
I have the same issue.  I have a Left side navigation for each part of a form.   As you navigate to a section, the form is shown for just that section of the object I'm trying to edit.  Imagine a json object like 
{ section1: { ...}, section2: {...} , etc } and a navigation pane for each section.  This is similar to the wizard example, but NOT the same because we want to validate ALL fields during sync validation, not just the current page.

**Comment by gustavohenke** [comment]
Because this issue saw almost no activity for a few months, I'm closing it.

**Comment by vitaliymaz** [comment]
Faced the same issue. I have a complicated form where some parts are hidden. And I have a general submit button which has to be disabled if the whole form is not valid. 
With current implementation I am not able to do that without any tricks. 
Are there some plans to fix this bug or I should just live with some workaround?

**Comment by ernani** [comment]
Maybe you can do this:

Validate each field individually or, create a general validation function
inside of your component (fired on every onChange evt for instance)

If you use the general function you can use `const selector =
formValueSelector('myComplexForm')` and pick the form values you need to be
appended in props object (`mapStateToProps`)

That general validate function will inspect `props.formValues` which
contains all form fields (as they get updated) so you can validate them on
every onChange method of each component.

You can then pass a `generalValidate('fieldName')` call to just validate
the specific field when it is triggering `onChange`, or, avoid passing the
field name to validate all fields existing in `props.formValues` as
`generalValidate() // no arguments`

Hope this helps!

On Wed, Dec 13, 2017 at 2:53 PM, Vitaliy Mazurenko <notifications@github.com

**Comment by fzaninotto** [comment]
Please reopen. This bug is still present, and it causes bugs in other libraries (e.g. https://github.com/marmelab/react-admin/issues/1866)

**Comment by danielrob** [comment]
While @gustavohenke closed this as having expired, it's an enhancement suggestion, not a bug. The existing behaviour is intended (e.g. Wizard form) and/or sufficient for most use cases. In the case of single page, not-always-rendered form fields/sections I'd recommend @psimoneau22's approach or similar. Since the field is not displayed, presumably it is only at submit time that the validation status actually matters. 

Basically, none of the existing maintainers are likely to fix this. If you feel it is an important case that redux-form should support and would like to contribute a PR this is warmly welcomed. First express your intent with how it would change the API, and the maintainers can give you the go ahead/or possible concerns, just to ensure that your time is not wasted.

**Comment by silenzium** [comment]
Even if the issue is old I'll share my solution/workaround for that problem.

I add an global `_error` to the error object if one of the unmounted fields is invalid. This way the form property `invalid` is true and my submit button gets disabled. Because the field is not rendered, I show the `_error` message above the submit button.

Perhaps this helps someone.

**Comment by kristian-oconnor** [comment]
An alternative option perhaps - is to not unmount the component that contain the fields for validation but hide it (using CSS or whatever) in the existing application, that way the fields are never unregistered... if they are never unregistered then the redux-form **selectors** (isDirty, isValid, isPristine etc) will work as expected as will the functions registered to 'reduxForm(...)' eg. onSubmit, onSubmitFail.

**Comment by terryc129** [comment]
I have experienced the same issue as @psimoneau22 , and tried different solutions. One option suggested from @kristian-oconnor to use css is not working for my case. For example, when I use Material UI, some component props will only be calculated once upon initial mount such as `labelWidth` prop in the TextField, if I hide & show the field by using css, `labelWidth` will be always `0` in this case.

When using field validation, I noticed the `syncErrors` in the redux state will capture the error even when the fields are unmounted. Thus I use `getFormSyncErrors` selector to get and validate the error before submitting the form. This solution works for me.
